### PR TITLE
Remove cas

### LINKED ISSUE #970: Remove CAS logic from file storage
As we plan our move to supporting Object Storage in core, removing existing CAS logic simplifies the transition to supporting both local and S3 style file storage. 

- [x] remove CAS from existing file storage
- [x] write migration script to create unique file for each file in mongo
- [ ] before upgrades, advise users to use existing one-off CAS calculation script to view how much extra space will be required after one-to-one file in mongo -> file object in storage is complete

### PR DESCRIPTION
**Pull request by davidfarkas**
Resolves #970

### Summary Added by @ryansanford
- Code will function with existing data in legacy CAS scheme
- New data will land in new non-CAS scheme
- Provided migration script
  - will copy, validate, db update, then remove file before moving onto the next one.
  - has no protection against parallel workers

### Review Checklist

- Tests were added to cover all code changes
- Documentation was added / updated
- Code and tests follow standards in CONTRIBUTING.md

### CODE REVIEW

File: api/placer.py [hunk 1]
```diff
@@ -494,7 +495,8 @@ def finalize(self):
             'filename': self.name,
             'path':	 self.path,
             'size':	 os.path.getsize(self.path),
-            'hash':	 files.hash_file_formatted(self.path),
+            'hash': files.hash_file_formatted(self.path),
+            'uuid':	 str(uuid.uuid4()),
```
Comment (nagem) [inline api/placer.py]: How do you feel about naming this field `_id` and using it as the `_id` field in Mongo when files become their own collection? Any reasons why they should be separate?

File: api/handlers/containerhandler.py [hunk 2]
```diff
@@ -111,7 +112,13 @@ def get(self, cont_name, **kwargs):
         # build and insert file paths if they are requested
         if self.is_true('paths'):
```
Comment (nagem) [inline api/handlers/containerhandler.py]: I don't believe this flag is used anywhere. I will confirm and hopefully we can just remove this path.

**Comment by ryansanford** [comment]
Added myself as reviewer to figure out how to approach testing this on dev integration system.

File: api/handlers/containerhandler.py (hunk 2 above)
Comment (nagem) [reply api/handlers/containerhandler.py]: Following up, support of the `paths` query param can be removed.

File: api/placer.py (hunk 1 above)
Comment (kofalt) [reply api/placer.py]: Agreed, let's follow Megan's convention if at all possible.

**Comment by ambrussimon** [comment]
@kofalt
1. `hash_file_formatted` vs object storage: it will indeed be different when we get to _uploading directly_ to storage, but that's further down the line. Also, I think getting a content hash can easily be part of the abstracted storage interface.
2. Agreed, "native" hashes :+1:

File: bin/oneoffs/remove_cas.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        os.makedirs(target_dir)
+    shutil.copy(path, target_path)
+
+
+def remove_cas():
+    """
+    Remove CAS logic, generate UUID for the files and rename them on the filesystem, make a copy of the file if more
+    than one container using the same hash.
+    """
+    COLLECTIONS_PREFIXES = [('projects', 'files'),
```
Comment (ryansanford) [inline bin/oneoffs/remove_cas.py]: @davidfarkas Are collection attachments (the scitran kind, not mongo) accounted for?

Comment (nagem) [reply bin/oneoffs/remove_cas.py]: Nice catch

**Comment by ryansanford** [comment]
@davidfarkas @ambrussimon Could either of you create me a script that will generate "placeholder" files in `SCITRAN_PERSISTENT_DATA_PATH` for all file references in the scitran db? Simple text file will suffice, where the contents is the full path, including file name of the file being created.

This will assist greatly with testing the upgrade using sanitized customer DBs without the corresponding data files.

Thanks!

**Comment by ryansanford** [comment]
Any objection to having the migrated/newly uploaded files land in a `v1` subfolder next to the existing `v0` one?

File: bin/oneoffs/remove_cas.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env bash
```
Comment (ryansanford) [inline bin/oneoffs/remove_cas.py]: This should be `#!/usr/bin/env python`, no?

**Comment by ryansanford** [comment]
With the migration script a file not existing should...
- not cause immediate exit
- still cause non-zero exit code
- print context to understand which project/session is impacted

```
2017-12-05 19:26:23      scitran.api                config.py   228:INFO Initializing database, creating indexes
2017-12-05 19:26:23      scitran.api                config.py   261:INFO Persisting configuration
2017-12-05 19:26:23       remove_cas            remove_cas.py    73:INFO copy file /var/scitran/data/v0/sha384/08/ea/v0-sha384-08ea14868e9c94b9011b4c51a64847d480599f141a796dc08079cef7883d6b1cb43b092a6b3952a9c11bd4d571767a15 to 2b/ae/2bae0b3f-4f42-4cab-8e4a-a4d6e845d501
2017-12-05 19:26:23       remove_cas            remove_cas.py    94:ERRO [Errno 2] No such file or directory: u'/var/scitran/data/v0/sha384/08/ea/v0-sha384-08ea14868e9c94b9011b4c51a64847d480599f141a796dc08079cef7883d6b1cb43b092a6b3952a9c11bd4d571767a15'
Traceback (most recent call last):
  File "bin/oneoffs/remove_cas.py", line 74, in remove_cas
    copy_file(f_path, os.path.join(base, util.path_from_uuid(f_uuid)))
  File "bin/oneoffs/remove_cas.py", line 32, in copy_file
    shutil.copy(path, target_path)
  File "/usr/lib/python2.7/shutil.py", line 119, in copy
    copyfile(src, dst)
  File "/usr/lib/python2.7/shutil.py", line 82, in copyfile
    with open(src, 'rb') as fsrc:
IOError: [Errno 2] No such file or directory: u'/var/scitran/data/v0/sha384/08/ea/v0-sha384-08ea14868e9c94b9011b4c51a64847d480599f141a796dc08079cef7883d6b1cb43b092a6b3952a9c11bd4d571767a15'
2017-12-05 19:26:23       remove_cas            remove_cas.py    95:INFO Rollback...
Traceback (most recent call last):
  File "bin/oneoffs/remove_cas.py", line 119, in <module>
    remove_cas()
  File "bin/oneoffs/remove_cas.py", line 100, in remove_cas
    uuid_path = util.path_from_uuid(f['_id'])
  File "/var/scitran/code/api/api/util.py", line 259, in path_from_uuid
    uuid_1 = uuid_.split('-')[0]
AttributeError: 'ObjectId' object has no attribute 'split'


```

**Comment by ryansanford** [comment]
Migration script should give some indication of relative progress. Maybe move existing info messages to debug, and put a less granular update for every 10,000 files similar to database.py?

File: bin/oneoffs/remove_cas.py (hunk 3 above)
Comment (ryansanford) [reply bin/oneoffs/remove_cas.py]: Just tested collection attachments, and they do get migrated. disregard

**Comment by ryansanford** [comment]
Phase 1 of my review is complete. 

I'd like the scripting requested (in this comment https://github.com/scitran/core/pull/989#issuecomment-349392494) for larger scale tests.

File: api/download.py [hunk 5]
```diff
@@ -47,7 +47,7 @@ def _append_targets(self, targets, cont_name, container, prefix, total_size, tot
                 if filtered:
                     continue
             if optional or not f.get('optional', False):
-                filepath = os.path.join(data_path, util.path_from_uuid(f.get('_id', '')))
+                filepath = os.path.join(data_path, 'v1', util.path_from_uuid(f.get('_id', '')))
```
Comment (kofalt) [inline api/download.py]: Would it be possible to move all these path joins to a helper function? There's 6+ instances of this logic, including the `v1` string.

**Comment by ryansanford** [comment]
@davidfarkas
Testing the migration with a Flywheel CLI uploaded gear... Looks like the file didn't get moved on the filesystem, and the Flywheel engine isn't able to pull down the gear rootfs due to a key error. This is not using the placeholder generation at all.


**API Log**
```
uwsgi_1           | 2017-12-11 15:26:06      scitran.api                  base.py    51:INFO Initialized request request_id=c9306fda-1513005966
uwsgi_1           | 2017-12-11 15:26:06      scitran.api                  base.py   373:ERRO Traceback (most recent call last):
uwsgi_1           |   File "/usr/local/lib/python2.7/dist-packages/webapp2.py", line 570, in dispatch
uwsgi_1           |     return method(*args, **kwargs)
uwsgi_1           |   File "./api/jobs/handlers.py", line 107, in download
uwsgi_1           |     file_id = gear['exchange']['rootfs-id']
uwsgi_1           | KeyError: 'rootfs-id'
uwsgi_1           |  request_id=c9306fda-1513005966
uwsgi_1           | [pid: 29|app: 0|req: 25/156] 172.19.0.8 () {32 vars in 426 bytes} [Mon Dec 11 15:26:06 2017] GET /api/gears/temp/5a2e9ff4fe061100174165ce => generated 101 bytes in 6 msecs (HTTP/1.1 500) 3 headers in 131 bytes (2 switches on core 0)
```

**Gear record after migration**
```
{
    "_id" : ObjectId("5a2e9ff4fe061100174165ce"),
    "category" : "converter",
    "gear" : {
        "inputs" : {
            "file" : {
                "base" : "file",
                "description" : "Any image."
            }
        },
        "maintainer" : "Ryan Sanford",
        "description" : "Gear created with gear builder",
        "license" : "Other",
        "author" : "Ryan Sanford",
        "url" : "http://example.example",
        "label" : "Ryan's Gear",
        "flywheel" : "0",
        "source" : "http://example.example",
        "version" : "0.0.1",
        "custom" : {
            "gear-builder" : {
                "image" : "flywheel/gear-base-anaconda"
            }
        },
        "config" : {},
        "name" : "ryans-gear"
    },
    "exchange" : {
        "git-commit" : "local",
        "rootfs-hash" : "sha384:9e30027893296b7bd5afa7d5edf4cc457a03ef335e4f77c482a085a088a30f58178fd3af458131f65d681d6123742a6e",
        "rootfs-url" : "/api/gears/temp/5a2e9ff4fe061100174165ce"
    },
    "created" : ISODate("2017-12-11T15:10:44.946Z"),
    "modified" : ISODate("2017-12-11T15:10:44.946Z")
}
```

**Comment by ryansanford** [comment]
@davidfarkas Ran into this after running the placeholder script, then the migration script. Guessing the placeholder script didn't add files for collection attachments?

Seems to be stuck at this point, no output for the last hour.

```
2017-12-11 17:02:07      scitran.api                config.py   228:INFO Initializing database, creating indexes
2017-12-11 17:02:07      scitran.api                config.py   261:INFO Persisting configuration
2017-12-11 17:02:08       remove_cas            remove_cas.py    95:INFO Processed 0 files of total 156174 files ...
2017-12-11 17:11:01       remove_cas            remove_cas.py    95:INFO Processed 15617 files of total 156174 files ...
2017-12-11 17:19:37       remove_cas            remove_cas.py    95:INFO Processed 31234 files of total 156174 files ...
2017-12-11 17:27:44       remove_cas            remove_cas.py    95:INFO Processed 46851 files of total 156174 files ...

2017-12-11 17:37:45       remove_cas            remove_cas.py    95:INFO Processed 62468 files of total 156174 files ...
2017-12-11 17:47:39       remove_cas            remove_cas.py    95:INFO Processed 78085 files of total 156174 files ...
2017-12-11 17:56:11       remove_cas            remove_cas.py    95:INFO Processed 93702 files of total 156174 files ...
2017-12-11 18:05:04       remove_cas            remove_cas.py    95:INFO Processed 109319 files of total 156174 files ...
2017-12-11 18:13:20       remove_cas            remove_cas.py    95:INFO Processed 124936 files of total 156174 files ...
2017-12-11 18:23:45       remove_cas            remove_cas.py    95:INFO Processed 140553 files of total 156174 files ...
2017-12-11 18:32:05       remove_cas            remove_cas.py    98:ERRO [Errno 2] No such file or directory: u'/var/scitran/data/v0/sha384/08/7e/v0-sha384-087e5402e2e1c305e68d11fe52d2e4af596dcd762b4b0f3383651b23243177ce510477039a76ddee6712df5ec491ce85'
Traceback (most recent call last):
  File "bin/oneoffs/remove_cas.py", line 73, in remove_cas
    copy_file(f_path, files.get_file_abs_path(f_uuid))
  File "bin/oneoffs/remove_cas.py", line 28, in copy_file
    shutil.copy(path, target_path)
  File "/usr/lib/python2.7/shutil.py", line 119, in copy
    copyfile(src, dst)
  File "/usr/lib/python2.7/shutil.py", line 82, in copyfile
    with open(src, 'rb') as fsrc:
IOError: [Errno 2] No such file or directory: u'/var/scitran/data/v0/sha384/08/7e/v0-sha384-087e5402e2e1c305e68d11fe52d2e4af596dcd762b4b0f3383651b23243177ce510477039a76ddee6712df5ec491ce85'
2017-12-11 18:32:05       remove_cas            remove_cas.py   103:ERRO (u"Wasn't able to migrate the 'search_result.json' file in the 'collections' collection (collection id: 5703ec35a90dac0027db6cbc)", IOError(2, 'No such file or directory'))
Traceback (most recent call last):
  File "bin/oneoffs/remove_cas.py", line 101, in remove_cas
    (f['fileinfo']['name'], f['collection'], str(f['collection_id'])), e)
MigrationError: (u"Wasn't able to migrate the 'search_result.json' file in the 'collections' collection (collection id: 5703ec35a90dac0027db6cbc)", IOError(2, 'No such file or directory'))
2017-12-11 18:32:05       remove_cas            remove_cas.py   104:INFO Rollback...

```

**Comment by davidfarkas** [comment]
@ryansanford Updated the create_placeholders script to add files for the collection attachments too, and now the remove_cas script shows the progress of the rollback too.
About the gear files. I need some time to fix that, probably it will be done by tomorrow.

**Comment by davidfarkas** [comment]
@ryansanford I updated the scripts to create placeholders for the gear files and migrate them.

**Pull request closed without merging**